specs/interop/predeploys.md

```diff
@@ -254,6 +288,77 @@ Note that the `relayMessage` function is `payable` to enable relayers to earn in
 To enable cross chain authorization patterns, both the `_sender` and the `_source` MUST be exposed via `public`
 getters.
 
+#### Sending Back Messages Hashes
+
+When sending back a message that failed to be relayed on the destination chain
+to the source chain, it's crucial to ensure the message can only be sent back
+to the `L2ToL2CrossDomainMessenger` contract in its source chain.
+
+This function has no auth, which allows anyone to send back a given message hash.
+The `RETURN_DELAY` variable is added to give the users enough time to replay their
+failed messages and to prevent malicious actors from performing a griefing attack
+by sending messages back immediately.
+
+Once the message is returned to the source chain, the message on the local chain is set
+as successful in the `successfulMessages` mapping to ensure non-replayability.
+
+```solidity
+function revert(bytes32 _messageHash) external nonReentrant {
+    if (successfulMessages[messageHash]) revert MessageAlreadyRelayed();
+
+    (uint256 messageTimestamp, uint256 messageSource) = failedMessages[messageHash];
+
+    // Return delay is necessary to avoid griefing attacks by providing users with time to
+    // replay their transactions give this function has no auth
+    if (block.timestamp <  messageTimestamp + RETURN_DELAY) revert DelayHasNotEnsued();
+
+    delete failedMessages[messageHash];
+
+    // This is set to true to ensure non-replayability on this chain
+    successfulMessages[messageHash] = true;
+
+    bytes memory data = abi.encodeCall(
+        L2ToL2CrossDomainMessenger.receiveMessageHash,
+        (messageSource, block.chainid, Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER, messageHash)
+    );
+    emit SentMessage(data);
+}
+```
+
+#### Storing Sent-back Message Hashes
+
+When receiving a sent-back message only message hashes
+of actual failed messages should be stored, for this we must ensure the origin,
+sender and caller are all the expected contracts.
+
+It's also important to ensure only the hashes of messages that were initiated
+in this chain are accepted and that they come from the correct destination
+chain id.
+
+If all checks have been successful, the message has is stored in the
+`returnedMessages` mapping. This enables smart contracts to read from it and
+check whether a message failed or not, and handle the failing case accordingly.
+
+```solidity
+function receiveMessageHash(uint256 _messageSource, uint256 _messageDestination, address _sender, bytes32 _messageHash) external {
```

Comment:
I would like to think about the naming here a bit more, just so its easy to explain and think about. There are 2 functions that need to be called in order. First you `revert`, then you `....`.

Some words that come to mind:
- `rollback`
- `revertFrom`
- `revertBy`
- `reverted`

Open to changing the `revert` function name, just would like both of them to be concise and play well together so its easy to understand conceptually what is going on. Perhaps `revert` and `reverted` (which has EVM connotations which isn't ideal) or `rollback` and `rolledback`  ? This implies a present and past tense with the language

Reply:
I think both revert and rollback have might be a bit misleading, since we are not changing any origin state back.
the function name should convey that a message wont be re-playable anymore, and that it's hash has been sent to it's origin domain.
naming is hard x.x

Reply:
I agree that naming is hard and I don't mean to nit, just would like cohesive/simple naming. Maybe we could use the interface:

- `expire(bytes32)`
- `expired(bytes32 hash, uint256 chainid)`

If we call the constant `EXPIRY_WINDOW` then it conveys a concept of the failed cross chain messages expiring if they fail and some amount of time passes. You would call `expire` on the remote domain which would then call `expired` on the local domain. I think you need the `chainid` to `require(chainid == block.chainid)` for security but need to double check to be sure

Reply:
`expire` is perfect, nice one!

Reply:
Added the changes with that nomenclature, I think is very clean now. The only thing we are missing is defining the value of `EXPIRY_WINDOW`. I added it to the constants table with a `TBD`. Is this something you may have more insight on? Too large of a value may be bad for UX. Too small of a value may not give enough time for the chain state to allow a message to be successfully relayed.

Also, do you think it makes sense for this to be constant, over it having a setter in case it needs to be adjusted?